File: src/main/java/org/hisp/dhis/rules/variables/ProgramRuleVariable.java
package org.hisp.dhis.rules.variables;

import org.hisp.dhis.antlr.ParserExceptionWithoutContext;
import org.hisp.dhis.rules.RuleExpression;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;

import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;

public class ProgramRuleVariable
    extends ScalarFunctionToEvaluate
{
    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        RuleVariableValue variableValue = visitor.getValueMap().get( ctx.programVariable().getText() );

        if ( variableValue == null )
        {
            throw new ParserExceptionWithoutContext( "Variable " + ctx.programVariable().getText() + " not present" );
        }

        return variableValue.value() == null ?
                variableValue.type().defaultValue() : getValueCastedByType( variableValue );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        String variable = ctx.programVariable().getText();

        if ( visitor.getItemStore().containsKey( variable ) )
        {
            visitor.getItemDescriptions().put( ctx.getText(), visitor.getItemStore().get( variable ).getDisplayName() );

            return visitor.getItemStore().get( ctx.programVariable().getText() ).getValueType().getValue();
        }

        throw new ParserExceptionWithoutContext(
            "Variable " + ctx.programVariable().getText() + " does not exist" );
    }
}

File: src/jvmMain/kotlin/org/dhis2/ruleengine/variables/ProgramRuleVariable.kt
package org.dhis2.ruleengine.variables

import org.dhis2.ruleengine.parser.expression.CommonExpressionVisitor
import org.dhis2.ruleengine.parser.expression.function.ScalarFunctionToEvaluate
import org.hisp.dhis.antlr.ParserExceptionWithoutContext
import org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext

class ProgramRuleVariable : ScalarFunctionToEvaluate() {
    override fun evaluate(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variableValue = visitor.getValueMap()[ctx.programVariable().text]
            ?: throw ParserExceptionWithoutContext("Variable " + ctx.programVariable().text + " not present")
        return variableValue.value ?: variableValue.ruleValueType.defaultValue
    }

    override fun getDescription(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variable = ctx.programVariable().text
        if (visitor.getItemStore().containsKey(variable)) {
            visitor.itemDescriptions[ctx.text] = visitor.getItemStore()[variable]!!.displayName
            return visitor.getItemStore()[ctx.programVariable().text]!!.valueType.value
        }
        throw ParserExceptionWithoutContext(
            "Variable " + ctx.programVariable().text + " does not exist"
        )
    }
}